core/src/main/java/com/linecorp/armeria/common/stream/HttpDeframer.java

```diff
@@ -0,0 +1,290 @@
+        final SubscriptionImpl subscriptionImpl = super.subscribe(subscription);
+        if (subscriptionImpl == subscription) {
+            eventLoop = subscription.executor();
+            deferredInit();
+        }
+        return subscriptionImpl;
+    }
+
+    private void deferredInit() {
+        final Subscription upstream = this.upstream;
+        if (upstream != null && eventLoop != null) {
+            if (initializedUpdater.compareAndSet(this, 0, 1)) {
+                if (demand() > 0) {
+                    upstream.request(1);
+                }
+            }
+        }
+    }
+
+    @Override
+    void request(long n) {
+        if (initialized != 0 && demand() == 0) {
+            upstream.request(1);
+        }
+        super.request(n);
+    }
+
+    @Override
+    void cancel() {
+        cancelAndCleanup();
+        super.cancel();
+    }
+
+    @Override
+    public void onSubscribe(Subscription subscription) {
+        requireNonNull(subscription, "subscription");
+        if (upstreamUpdater.compareAndSet(this, null, subscription)) {
+            deferredInit();
+        } else {
+            subscription.cancel();
+        }
+    }
+
+    @Override
+    public void onNext(HttpObject data) {
+        final EventExecutor eventLoop = this.eventLoop;
+        if (eventLoop.inEventLoop()) {
+            onNext0(data);
+        } else {
+            eventLoop.execute(() -> onNext0(data));
+        }
+    }
+
+    private void onNext0(HttpObject data) {
+        try {
+            process(data);
+
+            whenConsumed().thenRun(() -> {
+                if (demand() > 0) {
+                    upstream.request(1);
+                }
+            });
+        } catch (Throwable ex) {
+            handler.processOnError(ex);
+            cancelAndCleanup();
+            abort(ex);
+            Exceptions.throwIfFatal(ex);
+        }
+    }
+
+    private void cancelAndCleanup() {
+        cancelled = true;
+        final Subscription upstream = this.upstream;
+        if (upstream != null) {
+            upstream.cancel();
+        }
+        cleanup();
+    }
+
+    @Override
+    public void onError(Throwable cause) {
+        requireNonNull(cause, "cause");
+        if (cancelled) {
+            return;
+        }
+        final EventExecutor eventLoop = this.eventLoop;
+        if (eventLoop != null && !eventLoop.inEventLoop()) {
+            eventLoop.execute(() -> onError(cause));
+            return;
+        }
+
+        if (!(cause instanceof AbortedStreamException)) {
+            handler.processOnError(cause);
+        }
+
+        abort(cause);
+        cleanup();
+    }
+
+    @Override
+    public void onComplete() {
+        if (cancelled) {
+            return;
+        }
+
+        final EventExecutor eventLoop = this.eventLoop;
+        if (eventLoop != null && !eventLoop.inEventLoop()) {
+            eventLoop.execute(this::onComplete);
+            return;
+        }
```

Comment:
Can't we just remove the 277~ 280? because it the `eventLoop` is `null` `cleanup` and `close()` is called by another thread anyway?

Reply:
That's a good point. `cleanup` is not a thread safe. I think we need a volatile flag such as `cleaned`.